### ISSUE TITLE
Switch value isn't compiling with DDC

### ISSUE THREAD

**Issue by jodinathan** [dart-lang/sdk#51320]
simple `return switch (someEnum) { SomeEnum.foo => 'foo'};`

build_web_compilers 4.0.0
build_modules 5.0.0
Dart SDK version: 3.0.0-187.0.dev (dev) (Tue Jan 31 11:34:02 2023 -0800) on "macos_x64"

```
[SEVERE] build_web_compilers:ddc on lib/app_component.ddc.module:
Error compiling dartdevc module:pagebuilder_front|lib/app_component.ddc.js

Crash when compiling:
type 'ConstructorInvocation' is not a subtype of type 'ConstantExpression' of ' in type cast'

#0      CfeEnumOperations.getEnumElementValue (package:front_end/src/fasta/kernel/exhaustiveness.dart:162:59)
#1      EnumInfo._createEnumElements (package:_fe_analyzer_shared/src/exhaustiveness/static_types.dart:284:48)
#2      EnumInfo.enumElements (package:_fe_analyzer_shared/src/exhaustiveness/static_types.dart:271:45)
#3      EnumInfo.getEnumElement (package:_fe_analyzer_shared/src/exhaustiveness/static_types.dart:276:12)
#4      ExhaustivenessCache.getEnumElementStaticType (package:_fe_analyzer_shared/src/exhaustiveness/static_types.dart:214:36)
#5      convertConstantToSpace (package:front_end/src/fasta/kernel/exhaustiveness.dart:348:17)
#6      convertExpressionToSpace (package:front_end/src/fasta/kernel/exhaustiveness.dart:315:10)
#7      ExpressionCaseInfo.createSpace (package:front_end/src/fasta/kernel/exhaustiveness.dart:266:12)
#8      ConstantsTransformer._handleExhaustiveness (package:front_end/src/fasta/kernel/constant_evaluator.dart:806:28)
#9      ConstantsTransformer.visitSwitchStatement (package:front_end/src/fasta/kernel/constant_evaluator.dart:854:12)
#10     SwitchStatement.accept1 (package:kernel/ast.dart:10494:9)
#11     RemovingTransformer.transform (package:kernel/visitor.dart:1776:17)
#12     LabeledStatement.transformOrRemoveChildren (package:kernel/ast.dart:9985:16)
#13     RemovingTransformer.defaultTreeNode (package:kernel/visitor.dart:2068:10)
#14     TreeVisitor1.defaultStatement (package:kernel/visitor.dart:640:48)
#15     TreeVisitor1.visitLabeledStatement (package:kernel/visitor.dart:656:7)
#16     LabeledStatement.accept1 (package:kernel/ast.dart:9965:9)
#17     RemovingTransformer.transformOrRemove (package:kernel/visitor.dart:1812:21)
#18     RemovingTransformer.transformList (package:kernel/visitor.dart:2045:19)
#19     RemovingTransformer.transformStatementList (package:kernel/visitor.dart:1999:5)
#20     Block.transformOrRemoveChildren (package:kernel/ast.dart:9768:7)
#21     RemovingTransformer.defaultTreeNode (package:kernel/visitor.dart:2068:10)
#22     TreeVisitor1.defaultStatement (package:kernel/visitor.dart:640:48)
#23     TreeVisitor1.visitBlock (package:kernel/visitor.dart:645:38)
#24     ConstantsTransformer.visitBlock.<anonymous closure> (package:front_end/src/fasta/kernel/constant_evaluator.dart:790:27)
#25     ConstantsTransformer._handleExhaustiveness (package:front_end/src/fasta/kernel/constant_evaluator.dart:846:17)
#26     ConstantsTransformer.visitBlock (package:front_end/src/fasta/kernel/constant_evaluator.dart:789:12)
#27     Block.accept1 (package:kernel/ast.dart:9754:58)
#28     RemovingTransformer.transform (package:kernel/visitor.dart:1776:17)
#29     IfStatement.transformOrRemoveChildren (package:kernel/ast.dart:10731:16)
#30     RemovingTransformer.defaultTreeNode (package:kernel/visitor.dart:2068:10)
#31     TreeVisitor1.defaultStatement (package:kernel/visitor.dart:640:48)
#32     TreeVisitor1.visitIfStatement (package:kernel/visitor.dart:677:50)
#33     IfStatement.accept1 (package:kernel/ast.dart:10695:9)
#34     RemovingTransformer.transformOrRemove (package:kernel/visitor.dart:1812:21)
#35     RemovingTransformer.transformList (package:kernel/visitor.dart:2045:19)
#36     RemovingTransformer.transformStatementList (package:kernel/visitor.dart:1999:5)
#37     Block.transformOrRemoveChildren (package:kernel/ast.dart:9768:7)
#38     RemovingTransformer.defaultTreeNode (package:kernel/visitor.dart:2068:10)
#39     TreeVisitor1.defaultStatement (package:kernel/visitor.dart:640:48)
#40     TreeVisitor1.visitBlock (package:kernel/visitor.dart:645:38)
#41     ConstantsTransformer.visitBlock.<anonymous closure> (package:front_end/src/fasta/kernel/constant_evaluator.dart:790:27)
#42     ConstantsTransformer._handleExhaustiveness (package:front_end/src/fasta/kernel/constant_evaluator.dart:846:17)
#43     ConstantsTransformer.visitBlock (package:front_end/src/fasta/kernel/constant_evaluator.dart:789:12)
#44     Block.accept1 (package:kernel/ast.dart:9754:58)
#45     RemovingTransformer.transform (package:kernel/visitor.dart:1776:17)
#46     ConstantsTransformer.visitFunctionNode (package:front_end/src/fasta/kernel/constant_evaluator.dart:614:19)
#47     FunctionNode.accept1 (package:kernel/ast.dart:4182:9)
#48     RemovingTransformer.transform (package:kernel/visitor.dart:1776:17)
#49     ConstantsTransformer.visitProcedure.<anonymous closure> (package:front_end/src/fasta/kernel/constant_evaluator.dart:523:23)
#50     ConstantEvaluator.withNewEnvironment (package:front_end/src/fasta/kernel/constant_evaluator.dart:3996:18)
#51     ConstantsTransformer.visitProcedure (package:front_end/src/fasta/kernel/constant_evaluator.dart:521:23)
#52     Procedure.accept1 (package:kernel/ast.dart:3573:55)
#53     RemovingTransformer.transformOrRemove (package:kernel/visitor.dart:1812:21)
#54     RemovingTransformer.transformList (package:kernel/visitor.dart:2045:19)
#55     RemovingTransformer.transformProcedureList (package:kernel/visitor.dart:1927:5)
#56     ConstantsTransformer.visitClass.<anonymous closure> (package:front_end/src/fasta/kernel/constant_evaluator.dart:484:7)
#57     ConstantEvaluator.withNewEnvironment (package:front_end/src/fasta/kernel/constant_evaluator.dart:3996:18)
#58     ConstantsTransformer.visitClass (package:front_end/src/fasta/kernel/constant_evaluator.dart:479:23)
#59     Class.accept1 (package:kernel/ast.dart:1482:53)
#60     RemovingTransformer.transformOrRemove (package:kernel/visitor.dart:1812:21)
#61     RemovingTransformer.transformList (package:kernel/visitor.dart:2045:19)
#62     RemovingTransformer.transformClassList (package:kernel/visitor.dart:1891:5)
#63     ConstantsTransformer.convertLibrary (package:front_end/src/fasta/kernel/constant_evaluator.dart:441:5)
#64     transformLibraries (package:front_end/src/fasta/kernel/constant_evaluator.dart:129:26)
#65     KernelTarget.runBuildTransformations (package:front_end/src/fasta/kernel/kernel_target.dart:1534:19)
#66     KernelTarget.buildComponent.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:641:7)
<asynchronous suspension>
#67     withCrashReporting (package:front_end/src/fasta/crash.dart:136:12)
<asynchronous suspension>
#68     IncrementalCompiler.computeDelta.<anonymous closure> (package:front_end/src/fasta/incremental_compiler.dart:413:23)
<asynchronous suspension>
#69     _compile (package:dev_compiler/src/kernel/command.dart:337:37)
<asynchronous suspension>
#70     _CompilerWorker.performRequest (package:dev_compiler/ddc.dart:67:18)
<asynchronous suspension>
#71     AsyncWorkerLoop.run (package:bazel_worker/src/worker/async_worker_loop.dart:35:20)
<asynchronous suspension>
#72     internalMain (package:dev_compiler/ddc.dart:30:5)
<asynchronous suspension>
```

**Comment by nshahan** [comment]
@jodinathan I believe using `switch` as an expression is rolled up into the `patterns` experiment. Do you have any experiments enabled in your build?

If I flesh out your example code a little:

```dart
enum SomeEnum { foo, bar, baz }
void main() {
  var e = SomeEnum.foo;
  print(fn(e));
}

String fn(SomeEnum someEnum) {
  return switch (someEnum) { SomeEnum.foo => 'foo'};
}
```

This example works with DDC if you use `--enable-experiment=patterns`.

**Comment by jodinathan** [comment]
yes it is with patterns experiment @nshahan.

I changed the code to use old `switch return` style and the error still occurs. I guess it wasn't directly related to the `enum switch`.  
Unfortunately I can't easily repro because our project is very big and I can't seem to find the culprit from the stack trace.

Isn't it possible to add more information when an error occurs while compiling?

**Comment by nshahan** [comment]
@johnniwinther 
It looks like the failing cast is coming from the CFE from some code that was added pretty recently. 
https://github.com/dart-lang/sdk/commit/3184f2224bb36f48804cce1f983f57b6e7b23a97

**Comment by jodinathan** [comment]
@johnniwinther it still occurs 3.0.0-236.0.dev

**Comment by jodinathan** [comment]
it has been fixed with 3.0.0-263.0.dev
